src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs

```diff
@@ -167,43 +178,95 @@ private static SyntaxNode GetEnclosingCodeElementNode(Document document, SyntaxT
+        {
+            var doc = solution.GetDocument(commonLocation.SourceTree);
+            if (doc == null)
+            {
+                return null;
+            }
+
+            var document = solution.GetDocument(doc.Id);
             var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
             if (semanticModel == null)
             {
                 return null;
             }
 
-            cancellationToken.ThrowIfCancellationRequested();
+            var methodNode = FindEnclosingMethodNode(document, commonLocation.SourceSpan.Start);
 
-            var symbol = semanticModel.GetDeclaredSymbol(syntaxNode, cancellationToken);
-            if (symbol == null)
+            if (methodNode == null)
             {
                 return null;
             }
 
-            using (var progress = new CodeLensFindReferencesProgress(symbol, syntaxNode, searchCap: 0, cancellationToken: cancellationToken))
-            {
-                await SymbolFinder.FindReferencesAsync(symbol, solution, progress, null,
-                    progress.CancellationToken).ConfigureAwait(false);
+            var displayInfoService = document.GetLanguageService<ICodeLensDisplayInfoService>();
+            var fullName = displayInfoService.ConstructFullName(semanticModel, methodNode);
 
-                var referenceTasks = progress.Locations
-                    .Where(location => location.Kind != LocationKind.MetadataFile && location.Kind != LocationKind.None)
-                    .Distinct(LocationComparer.Instance)
-                    .Select(
-                        location =>
-                            GetDescriptorOfEnclosingSymbolAsync(solution, location, cancellationToken))
-                    .ToArray();
+            cancellationToken.ThrowIfCancellationRequested();
 
-                await Task.WhenAll(referenceTasks).ConfigureAwait(false);
+            return !string.IsNullOrEmpty(fullName) ? new ReferenceMethodDescriptor(fullName, document.FilePath) : null;
+        }
 
-                return referenceTasks.Select(task => task.Result);
-            }
+        public async Task<IEnumerable<ReferenceMethodDescriptor>> FindMethodReferenceLocationsAsync(Solution solution, DocumentId documentId, SyntaxNode syntaxNode, CancellationToken cancellationToken)
+        {
+            return await FindAsync(solution, documentId, syntaxNode,
+                async progress =>
+                {
+                    var descriptorTasks =
+                        progress.Locations
+                        .Where(location => location.Kind != LocationKind.MetadataFile && location.Kind != LocationKind.None)
+                        .Select(location =>
+                            TryCreateMethodDescriptorAsync(location, solution, cancellationToken))
+                        .ToArray();
+
+                    await Task.WhenAll(descriptorTasks).ConfigureAwait(false);
```

Comment:
WhenAll doesn't have an overload that takes a cancellation token, I guess it relies on the sub-tasks cancellation tokens?

---
In reply to: [74627996](https://github.com/dotnet/roslyn/pull/13109#discussion_r74627996) [](ancestors = 74627996)